File: frontend/acervo.ts
import axios from 'axios';
import { Observable } from './observable';
import { Cliente } from './Ccliente';
import { Bibliotecario } from './bibliotecario';
import { TipoLivro } from './tipo-livro';
import { Emprestimo } from './emprestimo';

export class Acervo extends Observable {
    private static instance: Acervo | null = null;
    private tipoLivros: TipoLivro[] = [];

    private constructor() {
        super();
    }

    static getInstance(): Acervo {
        if (!Acervo.instance) {
            Acervo.instance = new Acervo();
        }
        return Acervo.instance;
    }

    private _isBibliotecario(usuario: any): usuario is Bibliotecario {
        return usuario instanceof Bibliotecario;
    }

    async adicionarLivro(usuario: Bibliotecario, tipoLivro: TipoLivro, livro: any) {
        if (this._isBibliotecario(usuario)) {
            try {
                const response = await axios.post('http://localhost:3000/acervo/livro', {
                    usuario,
                    tipoLivro,
                    livro
                });
                this.notificarObservadores(response.data.mensagem);
            } catch (error: any) {
                console.error('Erro ao adicionar livro:', error.message);
            }
        } else {
            throw new Error("Operação permitida apenas para bibliotecários.");
        }
    }

    async removerLivro(usuario: Bibliotecario, idLivro: number) {
        if (this._isBibliotecario(usuario)) {
            try {
                const response = await axios.delete(`http://localhost:3000/acervo/livro/${idLivro}`);
                this.notificarObservadores(response.data.mensagem);
            } catch (error: any) {
                console.error('Erro ao remover livro:', error.message);
            }
        } else {
            throw new Error("Operação permitida apenas para bibliotecários.");
        }
    }

    async alugarLivro(cliente: Cliente, id: number) {
        if (cliente instanceof Cliente) {
            try {
                const response = await axios.post('http://localhost:3000/acervo/emprestimo', {
                    cpf: cliente.cpf,
                    livroId: id
                });
                const emprestimo = response.data.emprestimo;
                if (emprestimo) {
                    cliente.historicoEmprestimos.push(emprestimo);
                    this.notificarObservadores(`Livro ${emprestimo.livro.id} foi alugado`);
                } else {
                    console.log(`Livro ${id} não está disponível.`);
                }
                return emprestimo;
            } catch (error: any) {
                console.error('Erro ao alugar livro:', error.message);
            }
        }
    }

    async devolverLivro(cliente: Cliente, id: number) {
        if (cliente instanceof Cliente) {
            try {
                const response = await axios.post('http://localhost:3000/acervo/devolucao', {
                    cpf: cliente.cpf,
                    livroId: id
                });
                const mensagem = response.data.mensagem;
                if (mensagem.includes('devolvido')) {
                    const emprestimo = cliente.historicoEmprestimos.find(e => e.livro.id === id);
                    if (emprestimo) {
                        emprestimo.dataDevolucao = new Date();
                    }
                    this.notificarObservadores(mensagem);
                } else {
                    console.log(mensagem);
                }
            } catch (error: any) {
                console.error('Erro ao devolver livro:', error.message);
            }
        }
    }

    async visualizarLivros(): Promise<TipoLivro[]> {
        try {
            const response = await axios.get('http://localhost:3000/acervo');
            this.tipoLivros = response.data.tipoLivros || []; 
            return this.tipoLivros;
        } catch (error) {
            console.error('Erro ao carregar o acervo:');
            return [];  
        }
    }
    
}

File: src/biblioteca/Acervo.java
package biblioteca;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Acervo {
    private static Acervo instance = null;
    private List<TipoLivro> tipoLivros;
    private List<Separado> separados;

    private Acervo() {
        this.tipoLivros = new ArrayList<>();
    }

    public static Acervo getInstance() {
        if (instance == null) {
            instance = new Acervo();
        }
        return instance;
    }

    private boolean isBibliotecario(Usuario usuario) {
        return usuario instanceof Bibliotecario;
    }

    public void adicionarLivro(Bibliotecario bibliotecario, TipoLivro tipoLivro, Livro livro) {
        if (isBibliotecario(bibliotecario)) {
            tipoLivro.adicionar(livro);
            if (!tipoLivros.contains(tipoLivro)) {
                tipoLivros.add(tipoLivro); // Adiciona o TipoLivro ao acervo se ainda não estiver lá
            }
            System.out.println("Livro adicionado ao acervo.");
        } else {
            throw new IllegalArgumentException("Operação permitida apenas para bibliotecários.");
        }
    }

    public void removerLivro(Bibliotecario bibliotecario, int idLivro) {
        if (isBibliotecario(bibliotecario)) {
            for (TipoLivro tipoLivro : tipoLivros) {
                Livro livroRemovido = tipoLivro.removerPorId(idLivro);
                if (livroRemovido != null) {
                    System.out.println("Livro removido do acervo.");
                    return;
                }
            }
        } else {
            throw new IllegalArgumentException("Operação permitida apenas para bibliotecários.");
        }
    }

    public Emprestimo alugarLivro(Cliente cliente, int idLivro) {
        for (TipoLivro tipoLivro : tipoLivros) {
            for (Livro livro : tipoLivro.visualizar()) {
                if (livro.getId() == idLivro && livro.isDisponivel()) {
                    livro.setDisponivel(false);
                    Emprestimo emprestimo = new Emprestimo(cliente.getCpf(), new Date(), null, livro);
                    cliente.getHistoricoEmprestimos().add(emprestimo);
                    System.out.println("Livro " + idLivro + " foi alugado.");
                    return emprestimo;
                }
            }
        }
        System.out.println("Livro " + idLivro + " não está disponível.");
        return null;
    }

    public void devolverLivro(Cliente cliente, int idLivro) {
        for (Emprestimo emprestimo : cliente.getHistoricoEmprestimos()) {
            if (emprestimo.getLivro().getId() == idLivro) {
                
                emprestimo.setDataDevolucao(new Date());
                System.out.print("Livro " + idLivro + " foi devolvido,");
                for (TipoLivro tipoLivro : tipoLivros) {
                    for (Livro livro : tipoLivro.visualizar()) {
                        if (livro.getId() == idLivro){
                            if(tipoLivro.semreserva()){
                                emprestimo.getLivro().setDisponivel(true);
                                System.out.println("e está disponível.");
                            }else{
                                tipoLivro.notificarDevolucao(livro, instance);
                                System.out.println(" e ficará reservado.");
                                
                                
                            }

                        }

                    }

                }
                
                
                return;
            }
        }
    }

    public void cancelarSeparacao(Separado separacao){
        separados.remove(separacao);     
        System.out.println("o livro"+ separacao.getLivro().getId() +"não está mais separado para" + separacao.getCliente().getNome());  
    }

    public void separarLivro(Separado separacao){
        separados.add(separacao);
        System.out.println("o livro"+ separacao.getLivro().getId() +"não está mais separado para" + separacao.getCliente().getNome());
    }

    public List<TipoLivro> visualizarLivros() {
        return this.tipoLivros;
    }
}
